presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,169 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.sanity;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.UnionNode;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.QualifiedNameReference;
+import com.google.common.collect.ListMultimap;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.getExpressionTypes;
+import static com.facebook.presto.type.UnknownType.UNKNOWN;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Ensures that all the expressions and FunctionCall's matches their output symbols
+ */
+public final class TypeMatchingChecker
+        implements PlanSanityChecker.Checker
+{
+    public TypeMatchingChecker() {}
+
+    @Override
+    public void validate(Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types, PlanNode plan)
+    {
+        plan.accept(new Visitor(session, metadata, sqlParser, types), null);
+    }
+
+    private static class Visitor
+            extends SimplePlanVisitor<Void>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final SqlParser sqlParser;
+        private final Map<Symbol, Type> types;
+
+        public Visitor(Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.types = requireNonNull(types, "types is null");
+        }
+
+        @Override
+        public Void visitAggregation(AggregationNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            AggregationNode.Step step = node.getStep();
+
+            // Do checking only when the step is SINGLE since FINAL and PARTIAL aggregation might have intermediate representation types
+            if (step.equals(AggregationNode.Step.SINGLE)) {
```

Comment:
@raghavsethi  I tried it yesterday, but the FINAL aggregations also have the different signatures from the SINGLE one, as for the example of `avg(double):double`, the partial will be `avg(double):varchar` and the final will be `avg(varchar):double`, so I think we could only check for SINGLE aggregation.